### ISSUE TITLE
Heading should have default id

### ISSUE THREAD

**Issue by saurabhdaware** [razorpay/blade#904]
## Problem

Our `<Heading>Something</Heading>` in markup renders as `<h2>Something</h2>` but ideally it should have id so that that heading can be visited using hash URL.

## Solution

So Ideally it should have id with that text so that users can visit `https://example.com/page#something` to go to that heading
```html
<h2 id="something">Something</h2>
```

This is how github renders the headings - 
<img width="308" alt="image" src="https://user-images.githubusercontent.com/30949385/205873068-3ca86e1f-e742-4a05-915f-a73bc83d3dc9.png">

## Extra

In content-heavy pages, it is also nice enhancement to add link icon when someone hovers the heading so that the exact path to heading can be copy-pasted. If we can do this or have a prop to enable this, it will be nice enhancement.

<img width="276" alt="image" src="https://user-images.githubusercontent.com/30949385/205873633-b99ed5b8-3899-4d99-bc72-b006dde11c23.png">

**Comment by divyanshu013** [comment]
Don't know if adding `id` by default is a good idea because it can easily clash with an existing heading. Seems like a case for link component to me. Link as heading to be specific

**Comment by saurabhdaware** [comment]
Oh yes Link as heading is also a good idea!!

**Comment by anuraghazra** [comment]
We can also just expose the ID prop, users can manage the IDs if they want.

**Comment by saurabhdaware** [comment]
oh yes that makes more sense since this usecase will mostly only be there for docs

**Comment by divyanshu013** [comment]
I think we can go ahead and add the support for `id` prop on heading. Thoughts @saurabhdaware @anuraghazra @chaitanyadeorukhkar @kamleshchandnani ?

**Comment by saurabhdaware** [comment]
I think Kamlesh mentioned some concerns about this making it easier for consumers to go and just add any styling using css ```css
#someheading { color: red }
```

Another way would be to support some prop-like `isLink` that adds a tiny link icon somewhere and also adds id to heading internally. This way it will also make that experience consistent. 

<img width="303" alt="image" src="https://user-images.githubusercontent.com/30949385/234797169-27e819be-fd5d-49f9-92c5-2be319b23dcb.png">

**Comment by divyanshu013** [comment]
But even this way `id` would be accessible for CSS overrides:

```css
#with-overlay-props {
  color: red;
}
```

**Comment by neeraj719714** [comment]
This is a big usecase in Docs repo actually, Every Heading has to be a link with some hash-url. But the problem is we can't use the id prop since the content is written by Techwriters so we don't wan't them to use ID neither they are comfortable with it. Also we can't assign any random id because the link would look bad. It has to be meaningful.

We have solved this problem using Trie datastructure. During the build time we assign id attribute to each heading tag using it's `textContent`. And for uniqueness we use the trie to assign numbers for making it unique.

 In runtime it can be solved by using context api as well which can keep track of all the ids

**Comment by saurabhdaware** [comment]
I have started a discussion on this with our design team to see if its something we would like to standardise on blade's side or should we just enable `id` prop on Heading and let consumers handle this.

**Comment by saurabhdaware** [comment]
@neeraj719714 can you meanwhile share your current implementation? I'll take a look and see what all is needed

**Comment by saurabhdaware** [comment]
@neeraj719714 So we created this example here using one div that handles the hover styling of link, Link component, and Title / Heading component - https://codesandbox.io/s/heading-link-final-8m89kw?file=/src/Heading.tsx

This will also allow you to pass your own custom `id`. This can probably be created as a wrapper in docs repo right now since its limited to documentation right now.

We can explore if there are any other ways to support this in future or if this is something that needs to be standardised in design system itself.

**Comment by neeraj719714** [comment]
@saurabhdaware yes it should work if the user (of blade) is allowed to pass id prop. They will have to create a wrapper around Heading component and create their own heading component with it adding the logic for unique ids.

**Comment by neeraj719714** [comment]
also here is how we handle it in docs. [Link](https://github.com/razorpay/docs-website/blob/main/webpack-loaders/rehype-plugins/rehype-id-provider-plugin/IdProvider.js)

**Comment by saurabhdaware** [comment]
> @saurabhdaware yes it should work if the user (of blade) is allowed to pass id prop. They will have to create a wrapper around Heading component and create their own heading component with it adding the logic for unique ids.

Yup so we discussed this internally. The problem is, just supporting `id` won't solve the issue either. 

We want to hide / show a LinkIcon inside of Heading on hover of it. This requires children styling on hover of the item. Blade fundamentally doesn't supporting something like `<Heading id="xyz" _hover={{ LinkIcon: { display: 'block' } }}>MyHeading</Heading>` and it might take some time to conclude something like this. 

So for now it should be fine to have a wrapper because we'll need wrapper even if we support `id` right?
